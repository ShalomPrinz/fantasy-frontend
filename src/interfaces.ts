const roles = ["GK", "DEF", "MID", "ATT"] as const;
type PlayerRole = typeof roles[number];

interface Player {
  id: number;
  name: string;
  role: PlayerRole;
  team: string;
}

export { Player, PlayerRole };
