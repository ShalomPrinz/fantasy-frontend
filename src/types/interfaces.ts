import type { Jersey } from "../res";

const roles = ["GK", "DEF", "MID", "ATT"] as const;
type PlayerRole = typeof roles[number];

type Player = {
  id: number;
  name: string;
  role: PlayerRole;
  team: Jersey;
};

export { Player, PlayerRole };
