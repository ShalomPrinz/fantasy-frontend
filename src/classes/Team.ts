import { Player, PlayerRole } from "interfaces";

type Players = {
  [key in PlayerRole]?: Player[];
};

type PlayerPairs = [Players, Players];

class Team {
  readonly players: Players = {
    GK: [],
    DEF: [],
    MID: [],
    ATT: [],
  };
  count: number;

  constructor(players: Player[]) {
    this.count = 0;
    players.forEach((p) => {
      this.players[p.role]?.push(p);
      this.count += 1;
    });
  }

  byPairs(): PlayerPairs {
    return [
      {
        GK: this.players.GK,
        DEF: this.players.DEF,
      },
      {
        MID: this.players.MID,
        ATT: this.players.ATT,
      },
    ];
  }
}

export default Team;
export { PlayerPairs };
