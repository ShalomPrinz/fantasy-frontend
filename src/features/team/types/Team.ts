import { isPlayerRole } from "../../../types";
import type { Player, PlayerRole } from "../../../types";

type Players = {
  [key in PlayerRole]: Player[];
};

class Team {
  readonly players: Players = {
    GK: [],
    DEF: [],
    MID: [],
    ATT: [],
  };
  count: number;
  isUserTeam: boolean;

  constructor(players: Player[], isUserTeam?: boolean) {
    this.count = 0;
    this.isUserTeam = Boolean(isUserTeam);
    players?.forEach((p) => this.addPlayer(p));
  }

  addPlayer(p: Player) {
    if (!isPlayerRole(p.role) || this.contains(p)) return false;

    this.players[p.role].push(p);
    this.count += 1;
    return true;
  }

  removePlayer(p: Player) {
    if (!isPlayerRole(p.role) || !this.contains(p)) return false;

    const index = this.players[p.role].findIndex((pl) => pl.id === p.id);
    this.players[p.role].splice(index, 1);
    this.count -= 1;
    return true;
  }

  contains({ id, role }: Player) {
    if (!isPlayerRole(role)) return false;
    return this.players[role].findIndex((p) => p.id === id) !== -1;
  }

  clone() {
    const cloned = new Team([], this.isUserTeam);
    Object.values(this.players).forEach((arr) =>
      arr.forEach((p) => cloned.addPlayer(p))
    );
    return cloned;
  }
}

export default Team;
