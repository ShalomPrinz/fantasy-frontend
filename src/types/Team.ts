import { Player, PlayerRole } from "./interfaces";

type Players = {
  [key in PlayerRole]?: Player[];
};

type PlayerPairs = [Players, Players];

class Team {
  readonly players: Players = {
    GK: [],
    DEF: [],
    MID: [],
    ATT: [],
  };
  count: number;

  constructor(players: Player[]) {
    this.count = 0;
    players?.forEach((p) => this.addPlayer(p));
  }

  addPlayer(p: Player) {
    if (this.players[p.role] && !this.contains(p)) {
      this.players[p.role]!.push(p);
      this.count += 1;
      return true;
    } else return false;
  }

  removePlayer(p: Player) {
    if (this.players[p.role] && this.contains(p)) {
      const index = this.players[p.role]!.findIndex((pl) => pl.id === p.id);
      this.players[p.role]!.splice(index, 1);
      this.count -= 1;
      return true;
    } else return false;
  }

  contains({ id, role }: Player) {
    if (this.players[role]) {
      return this.players[role]!.findIndex((p) => p.id === id) !== -1;
    } else return false;
  }

  clone() {
    const cloned = new Team([]);
    Object.values(this.players).forEach((arr) =>
      arr.forEach((p) => cloned.addPlayer(p))
    );
    return cloned;
  }

  byPairs(): PlayerPairs {
    return [
      {
        GK: this.players.GK,
        DEF: this.players.DEF,
      },
      {
        MID: this.players.MID,
        ATT: this.players.ATT,
      },
    ];
  }
}

export default Team;
export { PlayerPairs };
