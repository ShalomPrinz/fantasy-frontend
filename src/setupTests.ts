// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import axios from "axios";
import { setupServer } from "msw/node";

import { handlers } from "../test";
import * as TeamContext from "./contexts/TeamContext";
import * as UserContext from "./features/authentication/contexts/UserContext";
import { LeagueInfo, Message, Player, Team, User } from "./types";

jest.setTimeout(10000);

axios.defaults.baseURL = "";

const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

interface MockedUserContext {
  logout?: () => Promise<void>;
  state?: UserContext.UserState;
  user?: User;
}

interface MockedUser {
  inbox?: Message[];
  leagues?: LeagueInfo[];
  name?: string;
  players?: Player[];
}

export const getTestUser = (user?: MockedUser): User => ({
  id: "id",
  inbox: user?.inbox || [],
  leagues: user?.leagues || [],
  name: user?.name || "",
  team: new Team(user?.players || []),
});

export const mockUser = (context: MockedUserContext) => {
  const userContext = jest.spyOn(UserContext, "useUser");

  const { logout, state, user } = context;
  const mockedLogout = typeof logout !== "undefined" ? logout : async () => {};
  const mockedState =
    typeof state !== "undefined" ? state : UserContext.UserState.LOGGED_USER;
  const mockedUser = typeof user !== "undefined" ? user : getTestUser();

  userContext.mockImplementation(() => ({
    loading: false,
    login: async () => {},
    logout: mockedLogout,
    register: async () => {},
    state: mockedState,
    user: mockedUser,
    invalidate: async () => {},
  }));

  return userContext.mockRestore;
};

export const mockTeam = (team: Team) => {
  const teamContext = jest.spyOn(TeamContext, "useTeamState");
  teamContext.mockImplementation(() => team);
};

export * from "../test";
