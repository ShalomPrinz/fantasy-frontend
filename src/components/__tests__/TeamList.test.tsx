import Team from "classes/Team";
import { Player } from "interfaces";
import { render } from "setupTests";

import { TeamList } from "../";

describe("TeamList", () => {
  it("should render TeamList component", () => {
    const players: Player[] = [
      { id: 0, name: "Ter Stegen", role: "GK", team: "Barcelona" },
    ];
    const team = new Team(players);

    const { asFragment } = render(<TeamList team={team} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
