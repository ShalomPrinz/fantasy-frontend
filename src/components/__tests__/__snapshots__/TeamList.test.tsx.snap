// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamList should render TeamList component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           GK 
        </h1>
        <div
          class="d-flex align-items-center ps-5"
        >
          <img
            class=""
            src="Barcelona.png"
            style="width: 60px;"
            title="Barcelona"
          />
          <h4
            class="text-truncate ps-2 my-0"
          >
             Ter Stegen 
          </h4>
        </div>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           DEF 
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           MID 
        </h1>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           ATT 
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
